### PR TITLE
Rewrite material dictionary.

### PR DESCRIPTION
**Pull request by MrIbby**
Replace item stack wrappers with predicates. This is much simpler and more customizable.

### CODE REVIEW

File: src/main/java/xyz/openmodloader/dictionary/Dictionaries.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package xyz.openmodloader.dictionary;
+
+public final class Dictionaries {
+
+    private Dictionaries() {}
+
+    public static final ItemStackDictionary MATERIALS = new ItemStackDictionary();
```
Comment (coolsquid) [inline src/main/java/xyz/openmodloader/dictionary/Dictionaries.java]: Needs JavaDocs. What is the material dictionary? How is it used? etc. Include a link to the material recipes.

File: src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        this.size = width * height;
+        this.matchers = matchers;
+        this.output = output;
+    }
+
+    /**
+     * Creates a new shaped material recipe
+     *
+     * @param output recipe output
+     * @param input  recipe input
```
Comment (coolsquid) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: The input has a rather special syntax. Describe it.

Comment (MrIbby) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: This is all I could think of: `size = width * height`

File: src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * Creates a new shaped material recipe
+     *
+     * @param width    recipe width
+     * @param height   recipe height
+     * @param matchers recipe input
+     * @param output   recipe output
+     */
+    public ShapedMaterialRecipe(int width, int height, List<Predicate<ItemStack>> matchers, ItemStack output) {
```
Comment (coolsquid) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: I'd change this to `public ShapedMaterialRecipe(int width, int height, ItemStack output, Predicate<ItemStack>... matchers)` for the sake of convenience.

File: src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java (hunk 2 above)
Comment (MrIbby) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: Nevermind

File: src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java (hunk 3 above)
Comment (MrIbby) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: I don't think that's a good idea. For modders, it's already confusing to have two constructors. If we change the list to varargs, people will most likely use it (incorrectly).

File: src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java (hunk 2 above)
Comment (MrIbby) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: For this reason, I added the link (to `CraftingManager`) below this line.

Comment (coolsquid) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: Not exactly what I was thinking of. You should describe what the array should consist of, the position of the elements, etc. A typical input for this, as I understand it, could be {"X X", " X ", 'X', "logWood"}, which isn't exactly easy to read from the code, especially not for beginners. You need to tell people what the input array should look like. Preferably with examples.

File: src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java (hunk 3 above)
Comment (coolsquid) [inline src/main/java/xyz/openmodloader/dictionary/ShapedMaterialRecipe.java]: The constructors can easily be distinguished by looking for the `int width, int height` parameters. As long as you include good javadocs, no one should get confused by it.

**Comment by phase** [comment]
This commit removes ItemStackWrappers. We're those merged? I thought that PR closed?

**Comment by MrIbby** [comment]
I closed that PR and opened this one. The functionality of item stack wrappers was replaced by predicates.

**Comment by phase** [comment]
I didn't know they ever went into the code base.

**Comment by MrIbby** [comment]
Yeah, @CoolSquid made the original material dictionary and recipes.

**Comment by phase** [comment]
I must have missed that :P

**Comment by coolsquid** [comment]
This is ready for merge. You just need to fix the conflict @MrIbby.

**Pull request merged**